core-plugins/src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java

```diff
@@ -181,15 +212,63 @@ public Schema getSchema() {
       return SCHEMA;
     }
 
+    public String extractPasswordFromUrl() {
+      int getLastIndexOfAtSign = path.lastIndexOf("@");
+      String authentication = path.substring(0, getLastIndexOfAtSign);
+      return authentication.substring(authentication.lastIndexOf(":") + 1);
+    }
+
+    public boolean authContainsSpecialCharacters() {
+      Pattern regularPasswordWithoutSpecialCharacters = Pattern.compile("[^A-Za-z0-9]");
+      Matcher regularPassword = regularPasswordWithoutSpecialCharacters.matcher(extractPasswordFromUrl());
+      return !regularPassword.matches();
+    }
+
     Map<String, String> getFileSystemProperties() {
-      if (fileSystemProperties == null) {
+      if (fileSystemProperties == null && !authContainsSpecialCharacters()) {
         return new HashMap<>();
       }
+      HashMap<String, String> fileSystemPropertiesMap = new HashMap<>();
+      if (fileSystemProperties != null) {
+        fileSystemPropertiesMap.putAll(GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE));
+      }
+
       try {
-        return GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE);
+        if (authContainsSpecialCharacters()) {
+          Path urlInfo;
+          String extractedPassword = extractPasswordFromUrl();
+          String encodedPassword = URLEncoder.encode(extractedPassword);
```

Comment:
why are we url encoding the password? seems like we just set the un-encoded password in filesystem property fs.(s)ftp.password.<host>.<user>

Reply:
We are encoding the password in order to avoid _IllegalArgumentExceptions_. 

If we don't encode the password, the Path (org.apache.hadoop.fs.Path) will throw an Exception and we won't be able to extract the necessary parameters from Path (host, port, scheme, username, password). 

In this way, we are sure that the Path will always be valid and we will be able to extract whatever we need to put the parameters in File System Properties.